File: go/atticprotocol/protocol.go
// Package atticprotocol implements the CLI text-based protocol for
// communication between CLI clients and the AtticServer emulator.
//
// This is the Go implementation of the protocol defined in the Swift
// AtticCore module. It is designed to be protocol-compatible while
// following Go idioms and best practices.
//
// Protocol Format:
//
//	Request (CLI -> Server):   CMD:<command> [arguments...]\n
//	Success Response:          OK:<response-data>\n
//	Error Response:            ERR:<error-message>\n
//	Async Event:               EVENT:<event-type> <data>\n
//	Multi-line Separator:      \x1E (Record Separator)
//
// Example Session:
//
//	CLI: CMD:ping
//	SRV: OK:pong
//	CLI: CMD:pause
//	SRV: OK:paused
//	CLI: CMD:read $0600 16
//	SRV: OK:data A9,00,8D,00,D4,A9,01,8D,01,D4,60,00,00,00,00,00
package atticprotocol

import (
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"time"
)

// Protocol constants matching the Swift implementation.
const (
	// CommandPrefix is the prefix for commands sent from CLI to Server.
	CommandPrefix = "CMD:"

	// OKPrefix is the prefix for success responses from Server to CLI.
	OKPrefix = "OK:"

	// ErrorPrefix is the prefix for error responses from Server to CLI.
	ErrorPrefix = "ERR:"

	// EventPrefix is the prefix for async events from Server to CLI.
	EventPrefix = "EVENT:"

	// MultiLineSeparator is the character used to separate multiple lines
	// in a single response (Record Separator character, ASCII 0x1E).
	MultiLineSeparator = "\x1E"

	// SocketPathPrefix is the prefix for server socket paths.
	SocketPathPrefix = "/tmp/attic-"

	// SocketPathSuffix is the suffix for server socket paths.
	SocketPathSuffix = ".sock"

	// MaxLineLength is the maximum allowed length for a protocol line in bytes.
	MaxLineLength = 4096

	// CommandTimeout is the default timeout for commands.
	CommandTimeout = 30 * time.Second

	// PingTimeout is the timeout used for ping commands during connection verification.
	PingTimeout = 1 * time.Second

	// ConnectionTimeout is the timeout for establishing connections.
	ConnectionTimeout = 5 * time.Second

	// ProtocolVersion is the version string for the CLI protocol.
	ProtocolVersion = "1.0"
)

// SocketPath returns the socket path for a given process ID.
func SocketPath(pid int) string {
	return fmt.Sprintf("%s%d%s", SocketPathPrefix, pid, SocketPathSuffix)
}

// CurrentSocketPath returns the socket path for the current process.
func CurrentSocketPath() string {
	return SocketPath(os.Getpid())
}

// DiscoverSockets finds all AtticServer sockets in /tmp.
// Returns socket paths sorted by modification time (most recent first).
func DiscoverSockets() ([]string, error) {
	pattern := filepath.Join("/tmp", "attic-*.sock")
	matches, err := filepath.Glob(pattern)
	if err != nil {
		return nil, fmt.Errorf("failed to glob sockets: %w", err)
	}

	// Get modification times and sort
	type socketInfo struct {
		path    string
		modTime time.Time
	}
	sockets := make([]socketInfo, 0, len(matches))

	for _, path := range matches {
		info, err := os.Stat(path)
		if err != nil {
			continue // Skip inaccessible sockets
		}
		sockets = append(sockets, socketInfo{
			path:    path,
			modTime: info.ModTime(),
		})
	}

	// Sort by modification time (most recent first)
	sort.Slice(sockets, func(i, j int) bool {
		return sockets[i].modTime.After(sockets[j].modTime)
	})

	result := make([]string, len(sockets))
	for i, s := range sockets {
		result[i] = s.path
	}

	return result, nil
}

// DiscoverSocket finds the most recently active AtticServer socket.
// Returns empty string if no socket is found.
func DiscoverSocket() string {
	sockets, err := DiscoverSockets()
	if err != nil || len(sockets) == 0 {
		return ""
	}
	return sockets[0]
}

File: Python/AtticCLI/attic_cli/protocol.py
"""Protocol constants and response parsing for the Attic CLI socket protocol.

These constants must match CLIProtocol.swift exactly. The protocol is frozen —
no changes unless accompanied by a major version bump.
"""

from dataclasses import dataclass


# --- Wire-format prefixes ---

COMMAND_PREFIX = "CMD:"
OK_PREFIX = "OK:"
ERROR_PREFIX = "ERR:"
EVENT_PREFIX = "EVENT:"

# Multi-line responses join lines with ASCII Record Separator (0x1E).
# The client splits on this character to recover individual lines.
MULTI_LINE_SEP = "\x1e"

# --- Socket path ---

SOCKET_PATH_PREFIX = "/tmp/attic-"
SOCKET_PATH_SUFFIX = ".sock"


def socket_path_for_pid(pid: int) -> str:
    """Return the expected socket path for a given AtticServer PID."""
    return f"{SOCKET_PATH_PREFIX}{pid}{SOCKET_PATH_SUFFIX}"


# --- Timeouts (seconds) ---

COMMAND_TIMEOUT = 30.0
PING_TIMEOUT = 1.0
CONNECTION_TIMEOUT = 5.0

# --- Buffer size ---

MAX_RECV = 4096

# --- Protocol version ---

PROTOCOL_VERSION = "1.0"


# --- Response parsing ---


@dataclass(frozen=True, slots=True)
class CLIResponse:
    """Parsed server response.

    Attributes:
        success: True for OK responses, False for errors.
        payload: The response data (after prefix stripping).
        is_multiline: True if payload contains MULTI_LINE_SEP characters.
    """

    success: bool
    payload: str
    is_multiline: bool

    @property
    def lines(self) -> list[str]:
        """Split a multi-line payload into individual lines."""
        if self.is_multiline:
            return self.payload.split(MULTI_LINE_SEP)
        return [self.payload] if self.payload else []


def parse_response(raw: str) -> CLIResponse:
    """Parse a raw protocol response line into a CLIResponse.

    Args:
        raw: A single response line from the server (newline already stripped).

    Returns:
        CLIResponse with success/error status and payload.

    Raises:
        ValueError: If the response doesn't match any known prefix.
    """
    if raw.startswith(OK_PREFIX):
        payload = raw[len(OK_PREFIX) :]
        return CLIResponse(
            success=True,
            payload=payload,
            is_multiline=MULTI_LINE_SEP in payload,
        )

    if raw.startswith(ERROR_PREFIX):
        return CLIResponse(
            success=False,
            payload=raw[len(ERROR_PREFIX) :],
            is_multiline=False,
        )

    raise ValueError(f"Unexpected response: {raw!r}")


@dataclass(frozen=True, slots=True)
class CLIEvent:
    """Parsed async event from the server.

    Attributes:
        kind: Event type — "breakpoint", "stopped", or "error".
        data: Raw event data string after the event type.
    """

    kind: str
    data: str


def parse_event(raw: str) -> CLIEvent:
    """Parse an EVENT: prefixed line into a CLIEvent.

    Args:
        raw: A line starting with EVENT: (prefix already verified by caller).

    Returns:
        CLIEvent with kind and data fields.
    """
    body = raw[len(EVENT_PREFIX) :]
    # Event format: "EVENT:<kind> <data>" or just "EVENT:<kind>"
    parts = body.split(" ", 1)
    kind = parts[0]
    data = parts[1] if len(parts) > 1 else ""
    return CLIEvent(kind=kind, data=data)
